crates/ruff_python_semantic/src/model.rs

```diff
@@ -662,12 +662,104 @@ impl<'a> SemanticModel<'a> {
                 }
             }
 
+            // FIXME: Shouldn't this happen above where `class_variables_visible` is set?
             seen_function |= scope.kind.is_function();
         }
 
         None
     }
 
+    /// Simulates a runtime load of a given [`ast::ExprName`].
+    ///
```

Comment:
It was very much unclear to me, until I spent quite a while reading through the code, what it meant to "simulate a runtime load" and why or how this is different from what `lookup_symbol` and `resolve_load` do.

It seems the differences are a) this method ignores typing-only bindings (this explains the name choice), and b) this method is supposed to run after all bindings have been collected, but still do a flow-sensitive lookup (i.e. find the applicable binding for a name at a particular point in control flow, which is not necessarily the end of the scope.) Or semi-flow-sensitive, anyway; flow-sensitive as approximated by "source order within the scope, with no understanding of branches."

Difference (a) makes sense, though taken alone it seems like it could be a flag rather than an entirely separate method.

I have a few questions/thoughts about difference (b), some of which may be due to lack of familiarity with Ruff and its semantic model.

It seems like this can be an improvement over Ruff's current model in some cases. But why is this PR the one introducing this capability? What is it about `TCH007` and `TCH008` that require this capability, when Ruff's many other rules involving looking up name bindings have so far gotten away without this?

My understanding is that largely Ruff handles this by doing name lookups _while constructing_ the semantic model, so inapplicable bindings just don't exist yet. Is that approach not workable in this case? Why do we need a method that must instead be run after all bindings are collected, but then ignores the "late" ones, in inner scopes?

If this capability isn't critical to the functioning of the new rules specifically, but is really a more general improvement to Ruff's semantic model, perhaps it should be looked at more holistically in view of all Ruff rules, and not introduced as part of this PR?

If we do definitely need this new method for these new rules to work at all, then I think this also should be more clearly described in a longer doc comment on the method.

Reply:
> My understanding is that largely Ruff handles this by doing name lookups while constructing the semantic model, so inapplicable bindings just don't exist yet. Is that approach not workable in this case? Why do we need a method that must instead be run after all bindings are collected, but then ignores the "late" ones, in inner scopes?

The main reason is that TCH007/TCH008 perform speculative lookups, rather than the ones that would actually happen at runtime, to ensure that we don't produce violations for the parts of the type alias value that need to available at runtime and vice-versa. (in fact they actually perform both the speculative and regular lookup to avoid violations/fixes going in circles)

The normal flow of the semantic model does not cover the speculative nature of questions like "what if this annotation wasn't a forward reference" or "what if this annotation was turned into a forward reference?".

There are certainly other ways to implement this check (i.e. performing the speculative lookup at the right time during semantic analysis), but this seemed like the least disruptive one (especially since at the time I wrote this there wasn't yet a cache for parsing string annotations) and matches my implementation in the flake8-plugin I wrote, so I can be more confident that it works and produces zero false positives in several large code-bases of mine.

I was also worried that speculatively walking the type definition earlier/later than it was supposed to could lead to other rules triggering or unwanted state leaking into the semantic model. It would be difficult to ensure that this excursion would have no unwanted side-effects.

> Difference (a) makes sense, though taken alone it seems like it could be a flag rather than an entirely separate method.

And in fact in my own implementation in the flake8 plugin they are the same method with a flag. However the potentially incorrect placement of the `seen_function` update prompted me to keep this method separate for now, in order to avoid potentially subtle regressions in `lookup_symbol`.

We could either take the risk and merge these methods now or put it off until later, when we're more confident that the method is stable.

Reply:
Actually I am wrong, I will almost always need to wait until all the bindings have been visited, since the outer scope lookups can be non source-order sensitive (e.g. looking up symbols in global scope from within a function). So there isn't really a way to perform the lookup I want to perform at the correct time, I will always need to wait until all the bindings have been visited. F405 also works this way, but can rely on the references caching which binding they point to, so it's easy to find the references that don't point anywhere.

We could do something similar here and keep a second state for every reference within a type definition (instead of just the one we use for `SemanticModel::resolve_name()`), one for the binding that will only be available at runtime, and one for the binding that will only be available at typing time. However, that seemed like overkill to me, especially since that information will not be useful for most rules. (It will however also be needed for the remaining TC1/TC2 rules that can remove quotes from annotations)

Reply:
These comments make sense to me! I'm satisfied that it's reasonable to add this method. I think it would be ideal if more of this context around what the method does and how it differs from the other two methods available for looking up names, were captured in its doc comment.

Otherwise, this PR looks OK to me.

Reply:
I updated the docstring, does this explanation make sense?